File: 3-Greedy Algorithms, Minimum Spanning Tree, and Dynamic Programming/1-Greedy Algorithm/4-Huffman Encoding/Huffman.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;

/**
 * Huffman encoding implementation.
 *
 * @author Ziang Lu
 */
public class Huffman {

    /**
     * Static nested Node class.
     */
    private static class Node implements Comparable<Node> {
        /**
         * Character stored in this Node.
         */
        private Character c;
        /**
         * Frequency of the character.
         */
        private int freq;
        /**
         * Reference to left child.
         */
        Node left;
        /**
         * Reference to right child.
         */
        Node right;

        /**
         * Constructor with parameter.
         * @param c character to store
         * @param freq frequency of the character
         */
        Node(Character c, int freq) {
            this(c, freq, null, null);
        }

        /**
         * Constructor with parameter.
         * @param c character to store
         * @param freq frequency of the character
         * @param left left child
         * @param right right child
         */
        Node(Character c, int freq, Node left, Node right) {
            this.c = c;
            this.freq = freq;
            this.left = left;
            this.right = right;
        }

        @Override
        public int compareTo(Node o) {
            return Integer.compare(freq, o.freq);
        }
    }

    /**
     * Root of the Huffman tree.
     */
    private Node huffmanTreeRoot;
    /**
     * Encoding map.
     */
    private HashMap<Character, String> encodingMap;

    /**
     * Constructor with parameter.
     * @param s string to build the Huffman tree
     */
    public Huffman(String s) {
//        constructHuffmanTreeWithHeap(s);
        constructHuffmanTreeWithTwoQueues(s);
    }

    /**
     * Constructs the Huffman tree using the given string with a heap.
     * @param s string to build the Huffman tree
     */
    public void constructHuffmanTreeWithHeap(String s) {
        // Check whether the input string is null or empty
        if ((s == null) || (s.length() == 0)) {
            huffmanTreeRoot = new Node(null, 0);
            return;
        }

        // Create a map between characters and their frequencies   [O(n)]
        HashMap<Character, Integer> freqOfChars = getFreqOfChars(s);

        // Create a Node heap from the map   [O(nlog n)]
        PriorityQueue<Node> heap = new PriorityQueue<Node>();
        for (Map.Entry<Character, Integer> entry : freqOfChars.entrySet()) {
            heap.offer(new Node(entry.getKey(), entry.getValue()));
        }

        // Construct the Huffman tree using a heap   [O(nlog n)]
        while (heap.size() > 1) {
            // Take out two nodes with minimum frequency from the heap
            Node minNode1 = heap.poll();
            Node minNode2 = heap.poll();
            // Combine the two nodes
            Node combined = new Node(null, minNode1.freq + minNode2.freq, minNode1, minNode2);
            // Put the combined node back to the heap
            heap.offer(combined);
        }
        // By now there is only one node in the heap, which is exactly the root of the Huffman tree.
        huffmanTreeRoot = heap.poll();

        // Create the encoding   [O(n)]
        encodingMap = new HashMap<Character, String>();
        createEncoding(huffmanTreeRoot, new StringBuilder());
        // Overall running time complexity: O(nlog n)
    }

    /**
     * Private helper method to create a map between the characters in the given
     * string and their frequencies.
     * @param s given string
     * @return map created map
     */
    private HashMap<Character, Integer> getFreqOfChars(String s) {
        HashMap<Character, Integer> freqOfChars = new HashMap<Character, Integer>();
        for (char c : s.toCharArray()) {
            Integer freq = freqOfChars.getOrDefault(c, 0);
            ++freq;
            freqOfChars.put(c, freq);
        }
        return freqOfChars;
        // Running time complexity: O(n)
    }

    /**
     * Private helper method to create the encoding recursively.
     * @param curr current node
     * @param encodingSoFar encoding so far
     */
    private void createEncoding(Node curr, StringBuilder encodingSoFar) {
        // Base case
        if ((curr.left == null) && (curr.right == null)) {
            encodingMap.put(curr.c, encodingSoFar.toString());
            return;
        }
        // Recursive case
        createEncoding(curr.left, encodingSoFar.append('0'));
        encodingSoFar.deleteCharAt(encodingSoFar.length() - 1); // Remember to remove the appended character back
        createEncoding(curr.right, encodingSoFar.append('1'));
        encodingSoFar.deleteCharAt(encodingSoFar.length() - 1); // Remember to remove the appended character back
        // T(n) = 2T(n/2) + O(1)
        // a = 2, b = 2, d = 0
        // According to Master Method, the running time complexity is O(n).
    }

    /**
     * Constructs the Huffman tree using the given string with two queues.
     * @param s string to build the Huffman tree
     */
    public void constructHuffmanTreeWithTwoQueues(String s) {
        // Check whether the input string is null or empty
        if ((s == null) || (s.length() == 0)) {
            huffmanTreeRoot = new Node(null, 0);
            return;
        }

        // Create a map between characters and their frequencies
        HashMap<Character, Integer> freqOfChars = getFreqOfChars(s);

        // Create a Node array and sort it   [O(nlog n)]
        ArrayList<Node> nodes = new ArrayList<Node>();
        for (Map.Entry<Character, Integer> entry : freqOfChars.entrySet()) {
            nodes.add(new Node(entry.getKey(), entry.getValue(), null, null));
        }
        Collections.sort(nodes);

        // Construct the Huffman tree using two queues   [O(n)]
        // "primitives" contains the nodes with the original character keys, while "merged" contains the merged nodes.
        // According to the implementation, both of the queues are kept sorted.
        ArrayDeque<Node> primitives = new ArrayDeque<Node>(nodes), merged = new ArrayDeque<Node>();
        while ((primitives.size() != 0) || (merged.size() != 1)) {
            // Take out two nodes with minimum frequency from the two queues
            Node minNode1 = getMinFreqNode(primitives, merged);
            Node minNode2 = getMinFreqNode(primitives, merged);
            // Combine the two nodes
            Node combined = new Node(null, minNode1.freq + minNode2.freq, minNode1, minNode2);
            // Put the combined node to the merged queue
            merged.offer(combined);
        }
        // By now there is only one node in the merged queue, which is exactly the root of the Huffman tree.
        huffmanTreeRoot = merged.pop();

        // Create the encoding [O(n)]
        encodingMap = new HashMap<Character, String>();
        createEncoding(huffmanTreeRoot, new StringBuilder());
        // Overall running time complexity: O(nlog n)
    }

    /**
     * Private helper method to get a node with minimum frequency from the given
     * two queues.
     * @param primitives queue of primitive nodes
     * @param merged queue of merged nodes
     * @return node with minimum frequency
     */
    private Node getMinFreqNode(ArrayDeque<Node> primitives, ArrayDeque<Node> merged) {
        if (primitives.size() == 0) {
            return merged.poll();
        } else if (merged.size() == 0) {
            return primitives.poll();
        }

        Node primitivesFront = primitives.peek();
        Node mergedFront = merged.peek();
        if (primitivesFront.compareTo(mergedFront) <= 0) {
            return primitives.pop();
        } else {
            return merged.pop();
        }
    }

    /**
     * Encodes the given message.
     * @param msg message to encode
     * @return encoded string
     */
    public String encode(String msg) {
        // Check whether the input string is null or empty
        if ((msg == null) || (msg.length() == 0)) {
            return "";
        }

        StringBuilder s = new StringBuilder();
        for (char c : msg.toCharArray()) {
            s.append(encodingMap.get(c));
        }
        return s.toString();
    }

    /**
     * Decodes the given encoded string.
     * @param encoded encoded string to decode
     * @return decoded message
     */
    public String decode(String encoded) {
        // Check whether the input string is null or empty
        if ((encoded == null) || (encoded.length() == 0)) {
            return "";
        }

        StringBuilder msg = new StringBuilder();
        decodeHelper(encoded, 0, huffmanTreeRoot, msg);
        return msg.toString();
    }

    /**
     * Private helper method to decode the encoded string recursively.
     * @param encoded encoded string to decode
     * @param idx index of the character in the encoded string to decode
     * @param curr current node
     * @param s decoded message
     */
    private void decodeHelper(String encoded, int idx, Node curr, StringBuilder s) {
        // Base case 1: Decoded all the encoded string
        if (idx >= encoded.length()) {
            s.append(curr.c);
            return;
        }
        // Base case 2: Reach a leaf
        if ((curr.left == null) && (curr.right == null)) {
            s.append(curr.c);
            // Restart from the root
            decodeHelper(encoded, idx, huffmanTreeRoot, s);
            return;
        }

        // Recursive case
        if (encoded.charAt(idx) == '0') {
            decodeHelper(encoded, idx + 1, curr.left, s);
        } else {
            decodeHelper(encoded, idx + 1, curr.right, s);
        }
    }

}

File: 3-Greedy Algorithms, Minimum Spanning Tree, and Dynamic Programming/1-Greedy Algorithm/5-Huffman Encoding/huffman.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Huffman encoding implementation.
"""

__author__ = 'Ziang Lu'

import heapq
from collections import Counter, deque
from functools import total_ordering
from typing import Deque


@total_ordering
class Node(object):
    __slots__ = ['_c', '_freq', '_left', '_right']

    def __init__(self, c: str, freq: int, left=None, right=None):
        """
        Constructor with parameter.
        :param c: str
        :param freq: int
        :param left: Node
        :param right: Node
        """
        self._c = c
        self._freq = freq
        self._left = left
        self._right = right

    @property
    def c(self) -> str:
        """
        Accessor of c.
        :return: str
        """
        return self._c

    @property
    def freq(self) -> int:
        """
        Accessor of freq.
        :return: int
        """
        return self._freq

    @property
    def left(self):
        """
        Accessor of left.
        :return: Node
        """
        return self._left

    @property
    def right(self):
        """
        Accessor of right.
        :return: None
        """
        return self._right

    def __lt__(self, other):
        return self._freq < other.freq


class HuffmanTree(object):
    __slots__ = ['_root', '_encoding_map']

    @staticmethod
    def _get_min_freq_node(primitives: Deque[Node],
                           merged: Deque[Node]) -> Node:
        """
        Gets a node with minimum frequency from the given two queues.
        :param primitives: deque[Node]
        :param merged: deque[Node]
        :return: Node
        """
        if not primitives:
            return merged.popleft()
        elif not merged:
            return primitives.popleft()

        primitives_front, merged_front = primitives[0], merged[0]
        if primitives_front <= merged_front:
            return primitives.popleft()
        else:
            return merged.popleft()

    def __init__(self, s: str):
        """
        Constructor with parameter.
        :param s: str
        """
        # self.construct_huffman_tree_with_heap(s)
        self.construct_huffman_tree_with_two_queues(s)

    def construct_huffman_tree_with_heap(self, s: str) -> None:
        """
        Constructs the Huffman tree using the given string with a heap.
        :param s: s
        :return: None
        """
        # Check whether the input string is None or empty
        if not s:
            self._root = Node(None, 0)
            return

        # Create a map between characters and their frequencies   [O(n)]
        freq_of_chars = Counter(s)

        # Create a Node heap from the map   [O(nlog n)]
        node_heap = []
        for c, freq in freq_of_chars.items():
            heapq.heappush(node_heap, Node(c, freq, left=None, right=None))

        # Construct the Huffman tree using a heap   [O(nlog n)]
        while len(node_heap) > 1:
            # Take out two nodes with minimum frequency from the heap
            min_node1 = heapq.heappop(node_heap)
            min_node2 = heapq.heappop(node_heap)
            # Combine the two nodes
            combined = Node(None, min_node1.freq + min_node2.freq,
                            left=min_node1, right=min_node2)
            # Put the combined node back to the heap
            heapq.heappush(node_heap, combined)
        # By now there is only one node in the heap, which is exactly the root
        # of the Huffman tree
        self._root = heapq.heappop(node_heap)

        # Create the encoding   [O(n)]
        self._encoding_map = {}
        self._create_encoding(self._root, encoding_so_far='')
        # Overall running time complexity: O(nlog n)

    def _create_encoding(self, curr: Node, encoding_so_far: str) -> None:
        """
        Private helper function to create the encoding recursively.
        :param curr: Node
        :param encoding_so_far: str
        :return: None
        """
        # Base case
        if not curr.left and not curr.right:
            self._encoding_map[curr.c] = encoding_so_far
            return
        # Recursive case
        self._create_encoding(curr.left, encoding_so_far=encoding_so_far + '0')
        self._create_encoding(curr.right, encoding_so_far=encoding_so_far + '1')
        # T(n) = 2T(n/2) + O(1)
        # a = 2, b = 2, d = 0
        # According to Master Method, the running time complexity is O(n).

    def construct_huffman_tree_with_two_queues(self, s: str) -> None:
        """
        Constructs the Huffman tree using the given string with two queues.
        :param s: str
        :return: None
        """
        # Check whether the input string is None or empty
        if not s:
            self._root = Node(None, 0)
            return

        # Create a map between characters and their frequencies   [O(n)]
        freq_of_chars = Counter(s)

        # Create a node array and sort it
        nodes = [
            Node(c, freq, left=None, right=None)
            for c, freq in freq_of_chars.items()
        ]
        nodes.sort()

        # Construct the Huffman tree using two queues   [O(n)]
        # "primitives" contains the nodes with the original character keys,
        # while "merged" contains the merged nodes.
        # According to the implementation, both of the queues are kept sorted.
        primitives, merged = deque(nodes), deque()
        while primitives or len(merged) != 1:
            # Take out two nodes with minimum frequency from the two queues
            min_node_1 = self._get_min_freq_node(primitives, merged)
            min_node_2 = self._get_min_freq_node(primitives, merged)
            # Combines the two nodes
            combined = Node(None, min_node_1.freq + min_node_2.freq,
                            left=min_node_1, right=min_node_2)
            # Put the combined node to the merged queue
            merged.append(combined)
        # By now there is only one node in the merged queue, which is exactly
        # the root of the Huffman tree.
        self._root = merged.popleft()

        # Create the encoding
        self._encoding_map = {}
        self._create_encoding(self._root, encoding_so_far='')

    def encode(self, msg: str) -> str:
        """
        Encodes the given message.
        :param msg: str
        :return: str
        """
        # Check whether the input string is None or empty
        if not msg:
            return ''

        return ''.join(map(lambda c: self._encoding_map[c], msg))

    def decode(self, encoded: str) -> str:
        """
        Decodes the given encoded string.
        :param encoded: str
        :return: str
        """
        # Check whether the input string is None or empty
        if not encoded:
            return ''

        return self._decode_helper(encoded, idx=0, curr=self._root)

    def _decode_helper(self, encoded: str, idx: int, curr: Node) -> str:
        """
        Private helper function to decode the encoded string recursively.
        :param encoded: str
        :param idx: int
        :param curr: Node
        :return: str
        """
        # Base case 1: Decoded all the encoded string
        if idx >= len(encoded):
            return curr.c
        # Base case 2: Reach a leaf
        if not curr.left and not curr.right:
            # Restart from the root
            return curr.c + \
                   self._decode_helper(encoded, idx=idx, curr=self._root)

        # Recursive case
        if encoded[idx] == '0':
            return self._decode_helper(encoded, idx=idx + 1, curr=curr.left)
        else:
            return self._decode_helper(encoded, idx=idx + 1, curr=curr.right)
